src/aeso_ast_infer_types.erl

```diff
@@ -1639,6 +1788,16 @@ infer_expr(Env, {app, Ann, Fun, Args0} = App) ->
             ResultType = fresh_uvar(Ann),
             when_warning(warn_unused_functions,
                          fun() -> register_function_call(Namespace ++ qname(CurrentFun), Name) end),
+            % the uvars of tvars are stored so that no variance switching happens
+            % in between them (e.g. in TC('a => 'a), 'a should be a single type)
+            case FunType of
```

Comment:
This is supposed to work in the case where the function is a type constructor. So it would work for something like:

```
let x = T(f)
```

where `f` is defined as:

```
function f(_ : Animal) : Cat = some_cat
```

and `T` is a type constructor for the datatype `tc`:

```
datatype tc('a) = T('a => 'a)
```

In the above example, when trying to infer the type for `T(f)`, the `FunType` for `T(f)` would look something like:

```
{fun_t,
    Ann1,
    [],
    [{fun_t, Ann2, [], [{uvar, Ann3, Ref}], {uvar, Ann4, Ref}}],
    {app_t, Ann5, {qid, Ann6, ["Main", "tc"]}, [{uvar, Ann7, Ref}]}}
```

Note that `Ref` in the return type is the same as the `Ref` from the args, since the type above is `    datatype tc('a) = TC('a => 'a)`, so it's enough to keep track of the `Ref` in the return type.